File: CDK/Sources/Network/CSURLRequest.cpp
//
//  CSURLRequest.cpp
//  CDK
//
//  Created by chan on 13. 6. 13..
//  Copyright (c) 2013년 brgames. All rights reserved.
//
#define CDK_IMPL

#include "CSURLRequest.h"

#include "CSURLRequestBridge.h"

#include "CSData.h"

CSURLRequest::~CSURLRequest() {
    CSURLRequestBridge::destroyHandle(_handle);
}

CSURLRequest* CSURLRequest::create(const char* url, const char* method, bool usingCache, float timeoutInterval) {
    void* handle = CSURLRequestBridge::createHandle(url, method, usingCache, timeoutInterval);
    
    return handle ? new CSURLRequest(handle) : NULL;
}

void CSURLRequest::setURL(const char* url) {
	CSURLRequestBridge::setURL(_handle, url);
}

const char* CSURLRequest::URL() const {
    return CSURLRequestBridge::URL(_handle);
}

void CSURLRequest::setMethod(const char* method) {
	CSURLRequestBridge::setMethod(_handle, method);
}

const char* CSURLRequest::method() const {
	return CSURLRequestBridge::method(_handle);
}

void CSURLRequest::setUsingCache(bool usingCache) {
	CSURLRequestBridge::setUsingCache(_handle, usingCache);
}

bool CSURLRequest::usingCache() const {
    return CSURLRequestBridge::usingCache(_handle);
}

void CSURLRequest::setTimeoutInterval(float timeoutInterval) {
	CSURLRequestBridge::setTimeoutInterval(_handle, timeoutInterval);
}

float CSURLRequest::timeoutInterval() const {
    return CSURLRequestBridge::timeoutInterval(_handle);
}

void CSURLRequest::setHeaderField(const char* name, const char* value) {
    CSURLRequestBridge::setHeaderField(_handle, name, value);
}

const char* CSURLRequest::headerField(const char* name) const {
	return CSURLRequestBridge::headerField(_handle, name);
}

const CSDictionary<CSString, CSString>* CSURLRequest::headerFields() const {
	return CSURLRequestBridge::headerFields(_handle);
}

void CSURLRequest::setBody(const CSData* data) {
    CSURLRequestBridge::setBody(_handle, data->bytes(), data->length());
}

void CSURLRequest::setBody(const void* data, uint length) {
    CSURLRequestBridge::setBody(_handle, data, length);
}

const CSData* CSURLRequest::body() const {
	return CSURLRequestBridge::body(_handle);
}

File: CDK/Sources/Platform/android/java/src/kr/co/brgames/cdk/CSURLRequest.java
package kr.co.brgames.cdk;

import android.util.Log;

import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;
import java.io.IOException;
import java.net.URL;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;

public class CSURLRequest {
    private float _timeoutInterval;
    private boolean _usingCache;
    private URL _url;
    private String _method;
    private final Map<String, String> _headers;
    private byte[] _body;
    
    private static final String TAG = "CSURLRequest";

    private CSURLRequest(String url, String method, boolean usingCache, float timeoutInterval) throws MalformedURLException {
        _url = new URL(url);
        _method = method;
        _usingCache = usingCache;
        _timeoutInterval = timeoutInterval;
        _headers = new HashMap<>();
    }
    
    public static CSURLRequest create(String url, String method, boolean usingCache, float timeout) {
        try {
            return new CSURLRequest(url, method, usingCache, timeout);
        }
        catch (MalformedURLException e) {
            Log.e(TAG, e.getMessage(), e);
            return null;
        }
    }

    public void setURL(String url) {
        try {
            _url = new URL(url);
        }
        catch (MalformedURLException e) {
            Log.e(TAG, e.getMessage(), e);
        }
    }

    public String URL() {
        return _url.toString();
    }

    public void setMethod(String method) {
        _method = method;
    }

    public String method() {
        return _method;
    }

    public void setHeaderField(String name, String value) {
        _headers.put(name, value);
    }

    public String[] headerFields() {
        String[] headers = new String[_headers.size() * 2];

        int i = 0;
        for (Map.Entry<String, String> header : _headers.entrySet()) {
            headers[i++] = header.getKey();
            headers[i++] = header.getValue();
        }
        return headers;
    }

    public Map<String, String> headers() {
        return _headers;
    }

    public void setBody(byte[] body) {
        _body = body;
    }

    public byte[] body() {
        return _body;
    }

    public void setUsingCache(boolean usingCache) {
        _usingCache = usingCache;
    }

    public boolean usingCache() {
        return _usingCache;
    }

    public void setTimeoutInterval(float timeoutInterval) {
        _timeoutInterval = timeoutInterval;
    }

    public float timeoutInterval() {
        return _timeoutInterval;
    }

    public HttpURLConnection openConnection() throws IOException {
        HttpURLConnection httpConn = (HttpURLConnection)_url.openConnection();

        httpConn.setUseCaches(_usingCache);
        int timeout = (int)(_timeoutInterval * 1000);
        httpConn.setConnectTimeout(timeout);
        httpConn.setReadTimeout(timeout);
        httpConn.setRequestMethod(_method);
        for (Map.Entry<String, String> header : _headers.entrySet()) {
            httpConn.setRequestProperty(header.getKey(), header.getValue());
        }
        if (_body != null) {
            httpConn.setDoOutput(true);
            httpConn.setFixedLengthStreamingMode(_body.length);
            try (OutputStream out = httpConn.getOutputStream()) {
                out.write(_body);
            }
        }
        return httpConn;
    }
}
